### PR TITLE
[Lam Kai Wen Jonathan] iP

### PR DESCRIPTION
**Pull request by lamkwjonathan**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,129 @@
+import java.util.Scanner;
+
 public class Duke {
+
+    public static final int END_OF_DONE_INDEX = 4;
+    public static final int END_OF_EVENT_INDEX = 6;
+    public static final int END_OF_DEADLINE_INDEX = 9;
+    public static final int END_OF_TODO_INDEX = 5;
```
Comment (sean-phang) [inline src/main/java/Duke.java]: I think it is good that you used named constants, avoiding magic numbers, making the code easier to understand

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,129 @@
...
+
+    private static void addTodoTask(String line) {
+        String description = line.substring(END_OF_TODO_INDEX);
+        tasks[numberOfTasks] = new Todo(description);
+        numberOfTasks++;
+    }
+
+    private static void exitProgram() {
+        isRunning = false;
     }
```
Comment (sean-phang) [inline src/main/java/Duke.java]: I think that there is a good level of abstraction being shown here

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,129 @@
...
+            } else if (line.startsWith("todo")) {
+                addTodoTask(line);
+                printTaskConfirmation();
+            } else if (line.startsWith("deadline")) {
+                addDeadlineTask(line);
+                printTaskConfirmation();
+            } else if (line.startsWith("event")) {
+                addEventTask(line);
+                printTaskConfirmation();
+            } else {
```
Comment (sean-phang) [inline src/main/java/Duke.java]: I think it is good that you used a default branch for the if-else statement, allowing you to detect errors.

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,129 @@
...
+        String description = line.substring(END_OF_EVENT_INDEX, endOfDescriptionIndex);
+        String at = line.substring(startOfAtIndex);
+
+        tasks[numberOfTasks] = new Event(description, at);
+        numberOfTasks++;
+    }
+
+    private static void addDeadlineTask(String line) {
+        int endOfDescriptionIndex = line.indexOf("/by") - 1;
+        int startOfByIndex = line.indexOf("/by") + 4;
```
Comment (sean-phang) [inline src/main/java/Duke.java]: Might be better to use a named constant for the number 4, to avoid magic numbers.

File: src/main/java/Task.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
Comment (sean-phang) [inline src/main/java/Task.java]: I think that it is good that you commented, however it might be repeating the obvious as the code is short and relatively self-explanatory.

File: src/main/java/Duke.java [hunk 6]
```diff
@@ -7,52 +7,87 @@ public static void main(String[] args) {
                 + "|  o  /|     ||  o  /|     |\n"
                 + "|  o  \\|  o  ||  o  \\|  o  |\n"
                 + "|_____/|_____||_____/|_____|\n";
-        String HORIZONTAL_LINE = "____________________________________________________________";
+        String horizontalLine = "____________________________________________________________";
```
Comment (leeyikai) [inline src/main/java/Duke.java]: For constants, I believe you can use the previous naming format which you used (HORIZONTAL_LINE)

File: src/main/java/Event.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+public class Event extends Task{
```
Comment (leeyikai) [inline src/main/java/Event.java]: Good job here! No coding violations found.

File: src/main/java/Duke.java [hunk 8]
```diff
@@ -7,52 +7,87 @@ public static void main(String[] args) {
...
-                System.out.println((taskNumber + 1) + ".["
-                        + Task.getTask(taskNumber).getStatusIcon() + "] "
-                        + Task.getTask(taskNumber).getDescription());
-                System.out.println(HORIZONTAL_LINE);
-            } else {
-                Task t = new Task(line);
+                System.out.println((taskNumber + 1) + "." + tasks[taskNumber]);
+                System.out.println(horizontalLine);
+            } else if (line.startsWith("todo")) {
+                String description = line.substring(5);
```
Comment (leeyikai) [inline src/main/java/Duke.java]: You can consider defining what is 5! For e.g. splitIndex = 5 (where the '/' is located)

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -7,52 +7,87 @@ public static void main(String[] args) {
...
 
+        Task[] tasks = new Task[100];
+        int numberOfTasks = 0;
         String line;
         boolean isRunning = true;
         Scanner in = new Scanner(System.in);
 
         System.out.println(Logo);
-        System.out.println(HORIZONTAL_LINE);
+        System.out.println(horizontalLine);
```
Comment (leeyikai) [inline src/main/java/Duke.java]: You can consider extracting this as a method! It might save you some time too

**Review by leeyikai — CHANGES_REQUESTED** [review]
Great job overall! minor feedback to increase code readability

File: src/main/java/duke/Duke.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    public static final int BY_LENGTH = 3;
+    public static final String LOGO = " _____         _____\n"
+            + "|     \\ _____ |     \\ _____\n"
+            + "|  o  /|     ||  o  /|     |\n"
+            + "|  o  \\|  o  ||  o  \\|  o  |\n"
+            + "|_____/|_____||_____/|_____|\n";
+    public static final String HORIZONTAL_LINE = "____________________________________________________________";
+    public static boolean isRunning = true;
+    public static Task[] tasks = new Task[100];
+    public static int numberOfTasks = 0;
```
Comment (ongweisheng) [inline src/main/java/duke/Duke.java]: Consider abstracting constants into a Constants class

File: src/main/java/duke/Duke.java [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+                } else if(line.trim().equals("list")) {
+                    printList();
+                } else if (line.startsWith("done")) {
+                    makeDone(line);
+                } else if (line.startsWith("todo")) {
+                    addTodoTask(line);
+                } else if (line.startsWith("deadline")) {
+                    addDeadlineTask(line);
+                } else if (line.startsWith("event")) {
+                    addEventTask(line);
```
Comment (ongweisheng) [inline src/main/java/duke/Duke.java]: Consider implementing a Parser class to parse user inputs

File: src/main/java/duke/Duke.java [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+    private static void sayBye() {
+        System.out.println(HORIZONTAL_LINE);
+        System.out.println("Ok. Bye bye!");
+        System.out.println(HORIZONTAL_LINE);
+    }
+
+    private static void sayInvalid() {
+        System.out.println(HORIZONTAL_LINE);
+        System.out.println("Sorry I don't understand that! Can you rephrase?");
+        System.out.println(HORIZONTAL_LINE);
```
Comment (ongweisheng) [inline src/main/java/duke/Duke.java]: Consider implementing a UI class to handle printing of outputs etc. Can see that some of your prints are repeated

File: src/main/java/duke/Duke.java [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+    private static void handleDone(String line) {
+        int taskNumber = Integer.parseInt(line.substring(END_OF_DONE_INDEX).trim()) - 1;
+        if (tasks[taskNumber].isDone()) {
+            System.out.println("This task is already done!");
+        } else {
+            tasks[taskNumber].markAsDone();
+            printDone(taskNumber);
+        }
+    }
```
Comment (ongweisheng) [inline src/main/java/duke/Duke.java]: Consider creating a TaskManager class to handle manipulation of tasks e.g. deadline, todo etc. Also the class can contain the list of tasks instead of it being at the start of main

File: src/main/java/duke/Duke.java [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+    private static void handleEventTask(String line) {
+        int endOfDescriptionIndex = line.indexOf("/at");
+        int startOfAtIndex = line.indexOf("/at") + AT_LENGTH;
+        String description = line.substring(END_OF_EVENT_INDEX, endOfDescriptionIndex).trim();
+        String at = line.substring(startOfAtIndex).trim();
+
+        tasks[numberOfTasks] = new Event(description, at);
+        numberOfTasks++;
+    }
```
Comment (ongweisheng) [inline src/main/java/duke/Duke.java]: Do remember to add javadoc comments down the line for more complicated methods that require some explanation

**Review by ongweisheng — COMMENTED** [review]
Generally well done as it seems like your code matches the requirements but do take note that you can apply more OOP apart from the different tasks categories
